packages/storage/src/AwsClients/S3/completeMultipartUpload.ts

```
@@ -0,0 +1,151 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import {
+	Endpoint,
+	getRetryDecider,
+	HttpRequest,
+	HttpResponse,
+	parseMetadata,
+} from '@aws-amplify/core/internals/aws-client-utils';
+import { composeServiceApi } from '@aws-amplify/core/internals/aws-client-utils/composers';
+import type {
+	CompleteMultipartUploadCommandInput,
+	CompleteMultipartUploadCommandOutput,
+	CompletedMultipartUpload,
+	CompletedPart,
+} from './types';
+import { defaultConfig } from './base';
+import {
+	map,
+	parseXmlBody,
+	parseXmlError,
+	s3TransferHandler,
+	serializeObjectSsecOptionsToHeaders,
+} from './utils';
+
+const INVALID_PARAMETER_ERROR_MSG =
+	'Invalid parameter for ComplteMultipartUpload API';
+
+export type CompleteMultipartUploadInput = Pick<
+	CompleteMultipartUploadCommandInput,
+	| 'Bucket'
+	| 'Key'
+	| 'UploadId'
+	| 'MultipartUpload'
+	| 'SSECustomerAlgorithm'
+	| 'SSECustomerKey'
+	| 'SSECustomerKeyMD5'
+>;
+
+export type CompleteMultipartUploadOutput = Pick<
+	CompleteMultipartUploadCommandOutput,
+	'$metadata' | 'Key' | 'ETag' | 'Location'
+>;
+
+const completeMultipartUploadSerializer = (
+	input: CompleteMultipartUploadInput,
+	endpoint: Endpoint
+): HttpRequest => {
+	const headers = serializeObjectSsecOptionsToHeaders(input);
+	const url = new URL(endpoint.url.toString());
+	url.hostname = `${input.Bucket}.${url.hostname}`;
+	url.pathname = `/${input.Key}`;
+	url.search = new URLSearchParams({ uploadId: input.UploadId }).toString();
+	return {
+		method: 'POST',
+		headers,
+		url,
+		body:
+			'<?xml version="1.0" encoding="UTF-8"?>' +
+			serializeCompletedMultipartUpload(input.MultipartUpload),
+	};
+};
+
+const serializeCompletedMultipartUpload = (
+	input: CompletedMultipartUpload
+): string =>
+	`<CompleteMultipartUpload xmlns="http://s3.amazonaws.com/doc/2006-03-01/">${(
+		input.Parts ?? []
```

I think it should throw `INVALID_PARAMETER_ERROR_MSG` here instead of falling back to an empty array, as empty array ~= invalid part data.